lib/babe/epoch_handler.go

```diff
@@ -0,0 +1,155 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package babe
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/ChainSafe/gossamer/lib/crypto/sr25519"
+)
+
+type handleSlotFunc = func(epoch, slotNum uint64, authorityIndex uint32, proof *VrfOutputAndProof) error
+
+var (
+	errEpochPast = errors.New("cannot run epoch that has already passed")
+)
+
+type epochHandler struct {
+	epochNumber uint64
+	firstSlot   uint64
+
+	constants *constants
+	epochData *epochData
+
+	// for slots where we are a producer, store the vrf output (bytes 0-32) + proof (bytes 32-96)
+	slotToProof map[uint64]*VrfOutputAndProof
+
+	handleSlot handleSlotFunc
+}
+
+func newEpochHandler(epochNumber, firstSlot uint64, epochData *epochData, constants *constants,
+	handleSlot handleSlotFunc, keypair *sr25519.Keypair) (*epochHandler, error) {
+	// determine which slots we'll be authoring in by pre-calculating VRF output
+	slotToProof := make(map[uint64]*VrfOutputAndProof)
+	for i := firstSlot; i < firstSlot+constants.epochLength; i++ {
+		proof, err := claimPrimarySlot(
+			epochData.randomness,
+			i,
+			epochNumber,
+			epochData.threshold,
+			keypair,
+		)
+		if errors.Is(err, errOverPrimarySlotThreshold) {
+			continue
+		} else if err != nil {
+			return nil, fmt.Errorf("error running slot lottery at slot %d: %w", i, err)
+		}
+
+		slotToProof[i] = proof
+		logger.Debugf("epoch %d: claimed slot %d", epochNumber, i)
+	}
+
+	return &epochHandler{
+		epochNumber: epochNumber,
+		firstSlot:   firstSlot,
+		constants:   constants,
+		epochData:   epochData,
+		slotToProof: slotToProof,
+		handleSlot:  handleSlot,
+	}, nil
+}
+
+func (h *epochHandler) run(ctx context.Context, errCh chan<- error) {
+	currSlot := getCurrentSlot(h.constants.slotDuration)
+
+	// if currSlot < h.firstSlot, it means we're at genesis and waiting for the first slot to arrive.
+	// we have to check it here to prevent int overflow.
+	if currSlot >= h.firstSlot && currSlot-h.firstSlot > h.constants.epochLength {
+		logger.Warnf("attempted to start epoch that has passed: current slot=%d, start slot of epoch=%d",
+			currSlot, h.firstSlot,
+		)
+		errCh <- errEpochPast
+		return
+	}
+
+	// for each slot we're handling, create a timer that will fire when it starts
+	// we create timers only for slots where we're authoring
+	authoringSlots := getAuthoringSlots(h.slotToProof)
+
+	type slotWithTimer struct {
+		timer   <-chan time.Time
```

Comment:
Use `*time.Timer` instead so you can stop them when the context gets canceled. You cannot cancel `time.After` and it will trigger even if the function has exit.